lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php

```diff
@@ -665,9 +666,26 @@ protected function prepareUpdateData($entity)
             // set $newVal = null, in order to insert a null value and schedule an
             // extra update on the UnitOfWork.
             if ($newValue !== null && $uow->isScheduledForInsert($newValue)) {
-                $uow->scheduleExtraUpdate($entity, [$propertyName => [null, $newValue]]);
+                // This is only required if the associated entity is different from the current one,
+                // and if the current entity relies on the database to generate its id
+                if ($newValue !== $entity) {
+                    $scheduleExtraUpdate = true;
+                } else {
+                    $identifiers = $this->class->getIdentifier();
+                    // Only single-column identifiers are supported
+                    if (1 === count($identifiers) && isset($entityChangeSet[$identifiers[0]])) {
```

Comment:
There should be an explicit method to ask if the identifier is single-valued? Otherwise, I'd suggest using the identifier flattener (it's a class lying around in internals)

Reply:
@Ocramius actually the identifier flattener returns an array too, not a hash as string.

And I've found any method to ask if the identifier is single-valued.